relay/pkg/header/integration_test.go

```diff
@@ -0,0 +1,292 @@
+		{Hash: to32Bytes(18), Height: 18, PrevHash: to32Bytes(17), Raw: toBytes(18)},
+		{Hash: to32Bytes(19), Height: 19, PrevHash: to32Bytes(18), Raw: toBytes(19)},
+		{Hash: to32Bytes(20), Height: 20, PrevHash: to32Bytes(19), Raw: toBytes(20)},
+	}
+
+	btcChain.SetHeaders(headers)
+
+	// Set the best known digest to header 7, so the relay will be pulling headers
+	// starting with the 8th header
+	localChain.SetBestKnownDigest(to32Bytes(7))
+
+	// Start the relay with difficulty changing at every eighth header and
+	// a shortened pushing loop sleep time
+	const (
+		testDifficultyEpochDuration = 8
+		testRelayPushingSleepTime   = 200 * time.Millisecond
+	)
+	startRelay(
+		ctx,
+		btcChain,
+		localChain,
+		testDifficultyEpochDuration,
+		relayPullingSleepTime,
+		testRelayPushingSleepTime,
+		&mockObserver{},
+	)
+	time.Sleep(10 * time.Millisecond)
+
+	//************ Verify processing of the first batch ************
+	addHeadersWithRetargetEvents := localChain.AddHeadersWithRetargetEvents()
+
+	expectedEventsCount := 1
+	actualEventsCount := len(addHeadersWithRetargetEvents)
+	if expectedEventsCount != actualEventsCount {
+		t.Fatalf(
+			"unexpected number of add headers with retarget events:\n"+
+				"expected: [%v]\n"+
+				"actual:   [%v]\n",
+			expectedEventsCount,
+			actualEventsCount,
+		)
+	}
+
+	expectedAddHeadersWithRetargetEvent := &chainlocal.AddHeadersWithRetargetEvent{
+		OldPeriodStartHeader: toBytes(0),
+		OldPeriodEndHeader:   toBytes(7),
+		Headers:              toBytes(8, 9, 10, 11, 12),
+	}
+
+	actualAddHeadersWithRetargetEvent := addHeadersWithRetargetEvents[0]
+
+	if !reflect.DeepEqual(
+		expectedAddHeadersWithRetargetEvent,
+		actualAddHeadersWithRetargetEvent,
+	) {
+		t.Errorf(
+			"unexpected add headers with retarget event:\n"+
+				"expected: [%+v]\n"+
+				"actual:   [%+v]\n",
+			expectedAddHeadersWithRetargetEvent,
+			actualAddHeadersWithRetargetEvent,
+		)
+	}
+
+	markNewHeaviestEvents := localChain.MarkNewHeaviestEvents()
+
+	expectedEventsCount = 1
+	actualEventsCount = len(markNewHeaviestEvents)
+	if expectedEventsCount != actualEventsCount {
+		t.Fatalf(
+			"unexpected number of mark new heaviest events:\n"+
+				"expected: [%v]\n"+
+				"actual:   [%v]\n",
+			expectedEventsCount,
+			actualEventsCount,
+		)
+	}
+
+	expectedMarkNewHeaviestEvent := &chainlocal.MarkNewHeaviestEvent{
+		AncestorDigest:    to32Bytes(7),
+		CurrentBestHeader: toBytes(7),
+		NewBestHeader:     toBytes(12),
+		Limit:             big.NewInt(6),
+	}
+	actualMarkNewHeaviestEvent := markNewHeaviestEvents[0]
+	if !reflect.DeepEqual(expectedMarkNewHeaviestEvent, actualMarkNewHeaviestEvent) {
+		t.Errorf(
+			"unexpected mark new heaviest event:\n"+
+				"expected: [%+v]\n"+
+				"actual:   [%+v]\n",
+			expectedMarkNewHeaviestEvent,
+			actualMarkNewHeaviestEvent,
+		)
+	}
+
+	localChain.SetBestKnownDigest(to32Bytes(12))
+	time.Sleep(testRelayPushingSleepTime)
+
+	//************ Verify processing of the second batch ************
```

Comment:
Can we add a short summary of what results are we expecting here in a human-friendly form?